compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ForLoopsLowering.kt

```diff
@@ -144,96 +173,140 @@ private class RangeLoopTransformer(
-        iteratorToLoopHeader[variable.symbol] = forLoopInfo
-
-        // Lower into a composite with additional declarations (e.g., induction variable)
-        // used in the loop condition and body.
+    private fun lowerHeader(variable: IrVariable, loopHeader: ForLoopHeader): IrStatement {
+        // Lower into a composite with additional statements (e.g., induction variable) used in the loop condition and body.
         return IrCompositeImpl(
             variable.startOffset,
             variable.endOffset,
             context.irBuiltIns.unitType,
             null,
-            forLoopInfo.declarations
+            loopHeader.loopInitStatements
         )
     }
 
-    override fun visitWhileLoop(loop: IrWhileLoop): IrExpression {
-        if (loop.origin != IrStatementOrigin.FOR_LOOP_INNER_WHILE) {
-            return super.visitWhileLoop(loop)
-        }
-
-        with(context.createIrBuilder(getScopeOwnerSymbol(), loop.startOffset, loop.endOffset)) {
-            // Visit the loop body to process the "next" statement and lower nested loops.
-            // Processing the "next" statement is necessary for building loops that need to
-            // reference the loop variable in the loop condition.
-            val newBody = loop.body?.transform(this@RangeLoopTransformer, null)?.let {
-                if (it is IrContainerExpression && !it.isTransparentScope) {
-                    IrCompositeImpl(startOffset, endOffset, it.type, it.origin, it.statements)
-                } else {
-                    it
-                }
-            }
-
-            val loopHeader = getLoopHeader(loop.condition)
-                ?: return super.visitWhileLoop(loop)  // If the for-loop cannot be lowered.
-            val (newLoop, replacementExpression) = loopHeader.buildLoop(this, loop, newBody)
-
-            // Update mapping from old to new loop so we can later update references in break/continue.
-            oldLoopToNewLoop[loop] = newLoop
-
-            return replacementExpression
-        }
-    }
+    private fun lowerWhileLoop(loop: IrWhileLoop, loopHeader: ForLoopHeader): LoopReplacement? {
+        val loopBodyStatements = (loop.body as? IrContainerExpression)?.statements ?: return null
+        val (mainLoopVariable, mainLoopVariableIndex, loopVariableComponents, loopVariableComponentIndices) = gatherLoopVariableInfo(
+            loopBodyStatements
+        )
 
-    private fun getLoopHeader(expression: IrExpression): ForLoopHeader? {
-        if (expression !is IrCall
-            || (expression.origin != IrStatementOrigin.FOR_LOOP_HAS_NEXT
-                    && expression.origin != IrStatementOrigin.FOR_LOOP_NEXT)
-        ) {
+        if (loopHeader.consumesLoopVariableComponents && mainLoopVariable.origin != IrDeclarationOrigin.IR_TEMPORARY_VARIABLE) {
+            // We determine if there is a destructuring declaration by checking if the main loop variable is temporary.
```

Comment:
Looks like we need more origins here to make such cases more explicit //cc @dnpetrov

Reply:
We could put it in a composite with `origin=DESTRUCTURING_DECLARATION`, that's how it works for regular destructuring statements (see `compiler/testData/ir/irText/expressions/destructuring1.txt`)